source/portable/NetworkInterface/DriverSAM/NetworkInterface.c

```diff
@@ -700,53 +701,72 @@ static BaseType_t prvGMACInit( NetworkInterface_t * pxInterface )
     /* Note that 'gmac_option.uc_copy_all_frame' is false, do not copy all frames.
      * And 'gmac_option.uc_no_boardcast' is false, meaning that broadcast is received.
      * 'boardcast' is a typo. */
-    memcpy( gmac_option.uc_mac_addr, pxEndPoint->xMACAddress.ucBytes, sizeof( gmac_option.uc_mac_addr ) );
+
+    /* Note that the gmac_option holds the MAC address that will be enabled for reception.
+     * Leave the MAC as zeros. It will be handled properly further down.*/
 
     gs_gmac_dev.p_hw = GMAC;
     gmac_dev_init( GMAC, &gs_gmac_dev, &gmac_option );
 
     NVIC_SetPriority( GMAC_IRQn, configMAC_INTERRUPT_PRIORITY );
     NVIC_EnableIRQ( GMAC_IRQn );
 
-    /* Clear the hash table for multicast MAC addresses.
-     * OR set both to ~0H to receive all multicast packets. */
-    GMAC->GMAC_HRB = 0U; /* Hash Register Bottom. */
-    GMAC->GMAC_HRT = 0U; /* Hash Register Top. */
+    /* The call to gmac_dev_init() above writes gmac_option.uc_mac_addr to the first
+     * MAC address register by calling gmac_set_address( GMAC, 0, p_opt->uc_mac_addr )
+     * This driver however uses the prvAddAllowedMACAddress() to manipulate the specific match
+     * and hash match registers, so undo the setting of the first specific MAC register. */
+
+    /* Disable all specific MAC address match registers */
+    for( int i = 0; i < GMACSA_NUMBER; i++ )
```

Comment:
I'd prefer if we do not use int. It is signed and the size is undefined. Maybe use a `size_t` or `uint32_t` or even `UBaseType_t`.
Or is there a specific reason that we want an int?

Reply:
I'd prefer size_t over all since `i` is used as an index of an array.

Reply:
I agree completely, but please help me out with the following:
In cases like that I regularly consider using UBaseType_t because it's supposed to be most efficient type for the platform, so it makes sense to use it, right? But then I should prefix the variable with 'ux' and then the very readable 'i' that a lot of programmers are used to should be called 'uxi'.
In essence I'm choosing readability because in cases like those for loops, the size and signedness doesn't really matter.
That's my reasoning. Please tell me how you feel about that silly "uxi" name. I'm out of town for a few days, but once I'm back I'll make those changes right away.

Reply:
Yes indeed, `uxi` will sound silly. No arguments there :)

I would not use `UBaseType_t` here though as this value is used as an index whose every possible value is supposed to fit in a `size_t` which may or may not be same as `UBaseType_t`.

To that end, I suggest we replace `i` with `xIndex`? I know that `for( int i ... )` is very familiar but indices should not be signed or they can be accidentally misused. In this case I know that we will never reach a value big enough which would cause the int to be negative, but it is a matter of consistency - which paves the path of secure and reliable code :)

Let me know if that makes sense.

Reply:
> I'd prefer if we do not use int. It is signed and the size is undefined. Maybe use a `size_t` or `uint32_t` or even `UBaseType_t`. Or is there a specific reason that we want an int?

The for-loop before C99 did not allow an inline declaration:
~~~diff
-    for( int i = 0; i < max; i++ ) {}
+    int i;
+    for( i = 0; i < max; i++ ) {}
~~~
MISRA doesn't like the use of `int`: `size_t` and eg. `uint32_t` are preferred.

The C Standard says that array indexes are (signed) integers. For a simple loop, my choice would be:
~~~
    BaseType_t xIndex;
    for( xIndex = 0; xIndex < max; xIndex++ ) {}
~~~
or when `max` is an unsigned number:
~~~
    size_t uxIndex;
    for( uxIndex = 0U; uxIndex < sizeof( xArray ); uxIndex++ ) {}
~~~

Reply:
@AniruddhaKanhere and @htibosch, Thank you both for sharing your opinion and most of all for explaining your reasoning. Please take a look at my latest commit and let me know if you like it better that way.
As far as the indexing, both of you had slightly different opinion, so I had to compromise. I went with `size_t xIndex`, but also remove one inline declaration from one of the for loops.
@AniruddhaKanhere, I consolidated the code that deals with the solicited-node addresses and MLD reports into a single function. It is a bit large and clunky but get's the job done of not repeating things like the IPv6 multicast address generation.
Let me know if you like it better this way.